beetsplug/fish.py

```diff
@@ -13,329 +14,393 @@
+            not_in_command = "not __fish_beet_subcommand"
+            script.complete(
+                f"(string split ' ' -- {names_text})",
+                conditions=[not_in_command],
+                description=command.help,
             )
 
-        if extravalues:
-            for f in extravalues:
-                setvar = wrap("$" + f.upper() + "S")
-                word += (
-                    " ".join(
-                        BL_EXTRA3.format(
-                            (cmdname + " " + f + ":"),
-                            ("-f " + "-A " + "-a " + setvar),
-                            ("-d " + wrap(f)),
-                        ).split()
-                    )
-                    + "\n"
-                )
-    return word
-
-
-def get_all_commands(beetcmds):
-    # Formatting for Fish to complete command options
-    word = ""
-    for cmd in beetcmds:
-        names = list(cmd.aliases)
-        names.append(cmd.name)
-        for name in names:
-            name = _escape(name)
-
-            word += "\n"
-            word += (
-                ("\n" * 2)
-                + "# ====== {} =====".format("completions for  " + name)
-                + "\n"
+        if include_fields:
+            # The set of fields to provide completions for.
+            fields = {
+                *library.Item.all_keys(),
+                *library.Album.all_keys(),
+                *extra_comp_fields,
+            }
+
+            # The completions include a ':' as it always follows.
+            field_comps = sorted(f"{f}:" for f in fields)
+            script.set_array("__fish_beet_flds", field_comps)
+            in_command = "__fish_beet_subcommand"
+            not_in_value = "not __fish_beet_metadata_param"
+            script.complete(
+                "$__fish_beet_flds",
+                conditions=[in_command, not_in_value],
+                description="known metadata field",
             )
 
-            for option in cmd.parser._get_all_options()[1:]:
-                cmd_l = (
-                    (" -l " + option._long_opts[0].replace("--", ""))
-                    if option._long_opts
-                    else ""
-                )
-                cmd_s = (
-                    (" -s " + option._short_opts[0].replace("-", ""))
-                    if option._short_opts
-                    else ""
-                )
-                cmd_need_arg = " -r " if option.nargs in [1] else ""
-                cmd_helpstr = (
-                    (" -d " + wrap(" ".join(option.help.split())))
-                    if option.help
-                    else ""
-                )
-                cmd_arglist = (
-                    (" -a " + wrap(" ".join(option.choices)))
-                    if option.choices
-                    else ""
+        if extra_comp_fields:
+            # The set of values for every user-specified extra field.
+            extra_values: Dict[str, set[str]] = dict.fromkeys(
```

Comment:
`from __future__ import annotations` may be missing at the top of the file

Reply:
That import would activate [PEP 563](https://peps.python.org/pep-0563/).  It's not necessary in this module, and even the Python 3.8 check doesn't complain in CI.  However, if we're instituting it as a convention across the codebase, we need some shared issue to point everybody to about it.

Reply:
![image](https://github.com/user-attachments/assets/67cd79ba-36e9-4c37-90f0-540305c80c67)

`mypy` complains about it since it cannot type-check it (this notation is only supported from Python 3.9 onwards). However, the CI is configured to ignore `mypy` errors.

This also fails **FA100** check from `flake8-future-annotations` linter, which nicely describes what it's about:

> # future-rewritable-type-annotation (FA100)
> 
> Derived from the **flake8-future-annotations** linter.
> 
> ## What it does
> Checks for missing `from __future__ import annotations` imports upon
> detecting type annotations that can be written more succinctly under
> PEP 563.
> 
> ## Why is this bad?
> PEP 585 enabled the use of a number of convenient type annotations, such as
> `list[str]` instead of `List[str]`. However, these annotations are only
> available on Python 3.9 and higher, _unless_ the `from __future__ import annotations`
> import is present.
> 
> Similarly, PEP 604 enabled the use of the `|` operator for unions, such as
> `str | None` instead of `Optional[str]`. However, these annotations are only
> available on Python 3.10 and higher, _unless_ the `from __future__ import annotations`
> import is present.
> 
> By adding the `__future__` import, the pyupgrade rules can automatically
> migrate existing code to use the new syntax, even for older Python versions.
> This rule thus pairs well with pyupgrade and with Ruff's pyupgrade rules.
> 
> This rule respects the [`target-version`] setting. For example, if your
> project targets Python 3.10 and above, adding `from __future__ import annotations`
> does not impact your ability to leverage PEP 604-style unions (e.g., to
> convert `Optional[str]` to `str | None`). As such, this rule will only
> flag such usages if your project targets Python 3.9 or below.
> 
> ## Example
> 
> ```python
> from typing import List, Dict, Optional
> 
> 
> def func(obj: Dict[str, Optional[int]]) -> None: ...
> ```
> 
> Use instead:
> 
> ```python
> from __future__ import annotations
> 
> from typing import List, Dict, Optional
> 
> 
> def func(obj: Dict[str, Optional[int]]) -> None: ...
> ```
> 
> After running the additional pyupgrade rules:
> 
> ```python
> from __future__ import annotations
> 
> 
> def func(obj: dict[str, int | None]) -> None: ...
> ```
> 
> ## Options
> - `target-version`

We have this import in every module that uses `list` / `dict` / `set` annotations.

Reply:
Thanks for the detailed explanation!  I've added the import.  However, I still have to use `typing.List` in one place: the explicit `cast(list[str], ...)` on line 299.  Since this expression is evaluated at runtime, Python 3.8 will fail with the standard "'type' object is not subscriptable" error.  As such, I haven't updated any of the types yet.

Reply:
Is that `cast` call required? 

Seems like `mypy` is happy with a simple annotation, and the plugin runs fine
![image](https://github.com/user-attachments/assets/a26b8774-7baf-41d1-b049-477c019dfbf7)

Reply:
Hm, I guess it's fine.  Will fix.